docs/guide/snaps-development-guide.md

```diff
@@ -1,47 +1,290 @@
+
+```javascript
+wallet.registerRpcMessageHandler(async (originString, requestObject) => {
+  switch (requestObject.method) {
+    case 'hello':
+      return 'world!';
+
+    default:
+      throw new Error('Method not found.');
+  }
+});
+```
+
+`wallet` is a global object that exposes the MetaMask Snap API, much like `window.ethereum` exposes our API to dapps.
+Any message sent via `wallet.request()` will be received and processed by MetaMask.
+In order to communicate with the outside world, the snap must implement its own RPC API by passing a handler function to `registerRpcMessageHandler`.
+Whenever the snap receives a JSON-RPC request from an external entity (a dapp or even another snap), the handler function will be called with the above parameters.
+
+If a dapp wanted to use `hello-snap`, it would do something like this:
+
+```javascript
+await ethereum.request({
+  method: 'wallet_enable',
+  params: [
+    {
+      wallet_snap: {
+        'npm:hello-snap': {},
+      },
+    },
+  ],
+});
+
+const hello = await ethereum.request({
+  method: 'wallet_invokeSnap',
+  params: ['npm:hello-snap', { method: 'hello' }],
+});
+
+console.log(hello); // 'world!'
+```
+
+The snap's RPC API is completely up to you, so long as it's a valid [JSON-RPC](https://www.jsonrpc.org/specification) API.
+
+::: tip Does my snap need to have an RPC API?
+Well, no, that's also up to you! If your snap can do something useful without receiving and responding to JSON-RPC requests, then you can skip calling `registerRpcMessageHandler`.
+However, if you want to do something like manage the user's keys for a particular protocol and create a dapp that e.g. sends transactions for that protocol via your snap, you need to specify an RPC API.
+:::
+
+### The Snap Manifest
+
+In order to get MetaMask to execute your snap, you need to have a valid manifest file, located in your package root directory under the name `snap.manifest.json`.
+The manifest file of `hello-snap` would look something like this:
+
+```json
+{
+  "version": "1.0.0",
+  "proposedName": "hello-snap",
+  "description": "A snap that says hello!",
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/Hello/hello-snap.git"
+  },
+  "source": {
+    "shasum": "w3FltkDjKQZiPwM+AThnmypt0OFF7hj4ycg/kxxv+nU=",
+    "location": {
+      "npm": {
+        "filePath": "dist/bundle.js",
+        "iconPath": "images/icon.svg",
+        "packageName": "hello-snap",
+        "registry": "https://registry.npmjs.org/"
+      }
+    }
+  },
+  "initialPermissions": {},
+  "manifestVersion": "0.1"
+}
+```
+
+The manifest tells MetaMask important information about your snap, most especially where it's published (via `source.location`) and how to verify the integrity of the snap source code (by attempting to reproduce the `source.shasum` value).
+For the time being, snaps can only be published to the official npm registry, and the manifest must also match the corresponding fields of the `package.json` file.
```

Comment:
I don't think we have an explanation of how to do this anywhere; obviously it's outside the scope of Snaps itself, but for example should we link to https://docs.npmjs.com/creating-and-publishing-scoped-public-packages ?

Reply:
I'll add a link to: https://docs.npmjs.com/packages-and-modules/contributing-packages-to-the-registry